Playbin: fix playback start glitch

xjuan opened the pull request:
https://phabricator.endlessm.com/T24734

cosimoc commented on src/playbin.js:
```diff
@@ -25,9 +26,22 @@ var Playbin = GObject.registerClass({
+
+        /* Update crop values on size allocation */
+        this.connect('size-allocate', (self, alloc) => {
+            self._updateCropArea(alloc);
```
You can directly call `this._updateCropArea()`

xjuan replied:
yeah I can, but then the linter will complain I did not used self right?

cosimoc replied:
I don't think so, but if it does, you can simply add `void self;` or something like that

ptomato replied:
if you use `alloc` then it will not complain about `self`

cosimoc replied:
Cool! You should be able to change this then

cosimoc commented on src/playbin.js:
```diff
@@ -44,41 +58,33 @@ var Playbin = GObject.registerClass({
 
-    _updateSizeFromStreamCollection (collection) {
+    _onStreamsSelected (msg) {
+        var collection = msg.parse_streams_selected();
```
Should this not be `const`?

ptomato replied:
I wonder why the linter in travis CI didn't catch this...

cosimoc commented on src/playbin.js:
```diff
@@ -102,29 +108,43 @@ var Playbin = GObject.registerClass({
+
+        if (newstate === Gst.State.PLAYING) {
+            this.set_visible_child(this._video_widget);
+            this.get_style_context().remove_class('locked');
```
I'd rather you set `locked = false` directly here, so there are not two places where style classes are tweaked... But also I would consider emitting a signal from here and setting it from the lockscreen through `_updateUI()` instead.

cosimoc commented on src/playbin.js:
```diff
@@ -142,22 +162,34 @@ playbin.locked {
         if (value)
             style.add_class('locked');
-        else
+        else if (!this.uri)
```
Why this check here? I'd rather move this check to whichever caller needed it, and keep this as a simple function.

ptomato commented on src/playbin.js:
```diff
@@ -1,22 +1,23 @@
         uri: GObject.ParamSpec.string('uri', 'URI', '',
             GObject.ParamFlags.READWRITE,
-            ''
+            null
```
JS doesn't have the concept of NULL strings, I'm not sure what will happen if you do this, but I think it might get marshalled as the empty string anyway, when setting the default value of the property. But I am pretty sure it's fine just to set `this._uri = null` in the `_init()` method as you do below, and leave this as-is.

ptomato commented on src/playbin.js:
```diff
@@ -25,9 +26,22 @@ var Playbin = GObject.registerClass({
+        });
+
+        /* Make button invisible */
+        this._button.connect('draw', () => {
```
`() => true` is shorter here

ptomato reviewed (commented):
Couple of drive-by comments

Commit:
Playbin: fix playback start glitch

Use a short transition animation on the stack when showing the video to make
sure the video widget is mapped before its visible.

Add 1 second transition to css background change to make adding a key while
the locked screen is visible smooth.

https://phabricator.endlessm.com/T24734

cosimoc commented on src/playbin.js:
```diff
@@ -1,33 +1,71 @@
+        super._init(props);
+
+        this.canDefault = false;
+        this.canFocus = false;
```
It would be nicer to fold these in the construction `props` dictionary above

cosimoc commented on src/playbin.js:
```diff
@@ -1,33 +1,71 @@
+        this.connect('draw', cr => {
+            const style = this.get_style_context();
+            const c = this.get_clip();
+            Gtk.render_background(style, cr, c.x, c.y, c.width, c.height);
```
How does this make the button invisible like the comment says? Also, I understand that code is here because the theme sets a background on the button, but buttons always draw a background in GTK, so you should not need the override...

cosimoc commented on src/playbin.js:
```diff
@@ -1,33 +1,71 @@
-            Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION
-        );
+        this.transitionType = Gtk.StackTransitionType.CROSSFADE;
+        this.transitionDuration = 100;
```
Please fold these into the `props` construction property dict.

cosimoc commented on src/playbin.js:
```diff
@@ -102,63 +132,81 @@ var Playbin = GObject.registerClass({
+        if (value) {
+            this._button.show();
+            this.set_visible_child(this._button);
+        } else if (!this._uri) {
```
I am not sure why this check needs to be here... Feels to me like it'd belong in whoever calls `locked = false`. I would like to keep this function "dumb" and just set the locked value it's passed, without state checks.
